improve kubectl cp command in several ways

WanLinghao opened the pull request:
**Release note**:
"kubectl cp" process soft link in better ways as well as some little bugs
  
**Soft link**:
before this patch 
    "kubectl cp" command will copy the soft link to destination as an empty regular file
after this patch
    "kubectl cp" command will behave the same as tar command
this patch improves it on both from container and to container

**some bugs**
1.from container to host
     a.when copy a file ends with '/', it will cause a panic.
        for example, container gakki has a regular file /tmp/test, then run command 
         _kubectl cp gakki:/tmp/test/  /tmp_
         a panic happens
 b.when copy a file which does not exist in container, the command ends up without 
         any error information

2.from host to container
    a.when run command like
        kubectl cp ""  gakki:/tmp
        it will try cp current directory to container, in other words, this command works
        the same as kubectl cp . gakki:/tmp
  b.current cp command will omit  an empty directory

mengqiy commented:
/assign @juanvallejo 
Assigning to you since you are working on kubectl cp as well.

mengqiy commented:
/ok-to-test

WanLinghao commented:
/test pull-kubernetes-e2e-gce-bazel 
/test pull-kubernetes-e2e-gce-etcd3
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-unit

juanvallejo commented on pkg/kubectl/cmd/cp.go:
```diff
@@ -244,22 +297,46 @@ func untarAll(reader io.Reader, destFile, prefix string) error {
+
+		mode := header.FileInfo().Mode()
+		if isFirstFile {
+			if !header.FileInfo().IsDir() && strings.HasSuffix(prefix, "/") {
```
while this is a great improvement over the current behavior, maybe we should consider removing the trailing slash from the filename in this case without returning the error.

cc @fabianofranz

WanLinghao replied:
thanks for the reply.
so you mean just copy the existing file without returning the error.
I refer the behaviour of cp a file from host to container.It will return an error with hint 
error:  not a directory.
so if your suggestion is better choice for cp command, I think it is necessary to keep cp command  in same pace  with both directions

WanLinghao replied:
@juanvallejo  are you still there? so it  supposed to return a error or just copy the file without any error?
which one is better?

juanvallejo replied:
Sorry about the delay.
At least with tools like `scp`, I'm able to specify a filename with a trailing slash, and have the command still work as expected.

I'd prefer copying the file without error, regardless of the trailing slash

WanLinghao replied:
OK, I will modify this and rebase this patch after your patch #51215 merged

WanLinghao commented:
/retest

WanLinghao committed:
This patch improve kubectl cp command from two aspects
A.support soft link better
before this patch
  "kubectl cp" command will copy the soft link to destination as an empty regular file
after this patch
  "kubectl cp" command will behave the same as tar command
  this patch improves it on both from container and to container

B.fix some bugs
1.from container to host
  a.when copy a file ends with '/', it will cause a panic.
    for example, container gakki has a regular file /tmp/test, then run command
    kubectl cp gakki:/tmp/test/ /tmp
    panic happens
  b.when copy a file which does not exist in container, the command ends up without
    any error information

2.from host to container
  a.when run command like
    kubectl cp "" gakki:/tmp
    it will try cp current directory to container, in other words, this command works
    the same as kubectl cp . gakki:/tmp
  b.current cp command will omit an empty directory

modified:   pkg/kubectl/cmd/cp.go
modified:   pkg/kubectl/cmd/cp_test.go

WanLinghao commented:
@juanvallejo rebased, PTAL, tks

juanvallejo commented on pkg/kubectl/cmd/cp.go:
```diff
@@ -222,7 +233,7 @@ func copyFromPod(f cmdutil.Factory, cmd *cobra.Command, cmderr io.Writer, src, d
 	}()
 	prefix := getPrefix(src.File)
-
+	prefix = path.Clean(prefix)
```
Why not just do `prefix := path.Clean(getPrefix(src.File))` in the line above?

juanvallejo commented:
@fabianofranz this lgtm

fabianofranz commented:
/sig cli
/release-label-none
/lgtm
/approve no-issue

CaoShuFeng commented:
/release-note-none

WanLinghao commented:
/release-note-none
